<p>Our Dataset is composed of mental health and frequency of mental health disorders in the tech workplace.</p>
<h2>Question to responde:</h2>
<h3>Taking treatment for a mental health condition?</h3>
<p>1. What are the most influencing features?</p>
<p>2. In this dataset is able to response if a person is talking a treatment or not?</p>
<h4>The Dataset:</h4>
<p>Size: 1259 rows e 27 features</p>
<ul>
    <li>Yes:50.595711</li>
    <li>No:49.404289</li>
</ul>
<p> The model is going to be train and an API in going to be available to test new cases.</p>
    


## Data set Mental Health in TI

<p>"This dataset is from a 2014 survey that measures attitudes towards mental health and frequency of mental health disorders in the tech workplace."(From Kanggle)</p>
<p> link: <a href="https://www.kaggle.com/datasets/osmi/mental-health-in-tech-survey?resource=download&select=survey.csv">Data source Kanggle</a></p>
<p>Version 2.0 - 12/10/2022</p>

In [1]:
## imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

## 1 Data Collection

In [2]:
# Lendo arquivo
df = pd.read_csv("data/survey.csv")

## 2 Data Understanding

In [3]:
df.shape

(1259, 27)

In [4]:
df.head()

Timestamp  Age  Gender         Country state self_employed  \
0  2014-08-27 11:29:31   37  Female   United States    IL           NaN   
1  2014-08-27 11:29:37   44       M   United States    IN           NaN   
2  2014-08-27 11:29:44   32    Male          Canada   NaN           NaN   
3  2014-08-27 11:29:46   31    Male  United Kingdom   NaN           NaN   
4  2014-08-27 11:30:22   31    Male   United States    TX           NaN   

  family_history treatment work_interfere    no_employees  ...  \
0             No       Yes          Often            6-25  ...   
1             No        No         Rarely  More than 1000  ...   
2             No        No         Rarely            6-25  ...   
3            Yes       Yes          Often          26-100  ...   
4             No        No          Never         100-500  ...   

                leave mental_health_consequence phys_health_consequence  \
0       Somewhat easy                        No                      No   
1          Don't know                     Maybe                      No   
2  Somewhat difficult                        No                      No   
3  Somewhat difficult                       Yes                     Yes   
4          Don't know                        No                      No   

      coworkers supervisor mental_health_interview phys_health_interview  \
0  Some of them        Yes                      No                 Maybe   
1            No         No                      No                    No   
2           Yes        Yes                     Yes                   Yes   
3  Some of them         No                   Maybe                 Maybe   
4  Some of them        Yes                     Yes                   Yes   

  mental_vs_physical obs_consequence comments  
0                Yes              No      NaN  
1         Don't know              No      NaN  
2                 No              No      NaN  
3                 No             Yes      NaN  
4         Don't know              No      NaN  

[5 rows x 27 columns]

In [5]:
#Tipos das colunas
df.dtypes

Timestamp                    object
Age                           int64
Gender                       object
Country                      object
state                        object
self_employed                object
family_history               object
treatment                    object
work_interfere               object
no_employees                 object
remote_work                  object
tech_company                 object
benefits                     object
care_options                 object
wellness_program             object
seek_help                    object
anonymity                    object
leave                        object
mental_health_consequence    object
phys_health_consequence      object
coworkers                    object
supervisor                   object
mental_health_interview      object
phys_health_interview        object
mental_vs_physical           object
obs_consequence              object
comments                     object
dtype: object

In [6]:
#Soma nulos por Coluna
df.isnull().sum()

Timestamp                       0
Age                             0
Gender                          0
Country                         0
state                         515
self_employed                  18
family_history                  0
treatment                       0
work_interfere                264
no_employees                    0
remote_work                     0
tech_company                    0
benefits                        0
care_options                    0
wellness_program                0
seek_help                       0
anonymity                       0
leave                           0
mental_health_consequence       0
phys_health_consequence         0
coworkers                       0
supervisor                      0
mental_health_interview         0
phys_health_interview           0
mental_vs_physical              0
obs_consequence                 0
comments                     1095
dtype: int64

In [7]:
# Soma todos os NaN do dataset
df.isna().sum().sum()

1892

## 3 Data Preparation

### Dealing with missing values
<p>Missing values in columns</p>
<ul>
<li>state - 515</li>
<li>self_employed - 18</li>
<li>work_interfere - 264</li>
<li>comments - 1095</li>
</ul>



#### Dropping Columns

In [8]:
# Drop column: comments
df.drop(["comments"], axis=1, inplace=True)
# Drop TimeStamp
df.drop(["Timestamp"], axis=1, inplace=True)

In [9]:
df.columns

Index(['Age', 'Gender', 'Country', 'state', 'self_employed', 'family_history',
       'treatment', 'work_interfere', 'no_employees', 'remote_work',
       'tech_company', 'benefits', 'care_options', 'wellness_program',
       'seek_help', 'anonymity', 'leave', 'mental_health_consequence',
       'phys_health_consequence', 'coworkers', 'supervisor',
       'mental_health_interview', 'phys_health_interview',
       'mental_vs_physical', 'obs_consequence'],
      dtype='object')

In [10]:
# Sum of null values
df.isnull().sum()

Age                            0
Gender                         0
Country                        0
state                        515
self_employed                 18
family_history                 0
treatment                      0
work_interfere               264
no_employees                   0
remote_work                    0
tech_company                   0
benefits                       0
care_options                   0
wellness_program               0
seek_help                      0
anonymity                      0
leave                          0
mental_health_consequence      0
phys_health_consequence        0
coworkers                      0
supervisor                     0
mental_health_interview        0
phys_health_interview          0
mental_vs_physical             0
obs_consequence                0
dtype: int64

#### Replacing with null values
<ul>
<li>self_employed - 18</li>
<li>work_interfere - 264</li>
<li>state - 515</li>
</ul>

In [11]:
df['self_employed'].value_counts(dropna=False)

No     1095
Yes     146
NaN      18
Name: self_employed, dtype: int64

In [12]:
df['work_interfere'].value_counts(dropna=False)

Sometimes    465
NaN          264
Never        213
Rarely       173
Often        144
Name: work_interfere, dtype: int64

In [13]:
df['state'].value_counts(dropna=False)

NaN    515
CA     138
WA      70
NY      57
TN      45
TX      44
OH      30
IL      29
OR      29
PA      29
IN      27
MI      22
MN      21
MA      20
FL      15
NC      14
VA      14
WI      12
GA      12
MO      12
UT      11
CO       9
MD       8
AL       8
AZ       7
OK       6
NJ       6
KY       5
SC       5
IA       4
CT       4
DC       4
NV       3
VT       3
SD       3
KS       3
NH       3
WY       2
NM       2
NE       2
WV       1
ID       1
MS       1
RI       1
LA       1
ME       1
Name: state, dtype: int64

#### Interpole data to self_employed and work_interfere

In [14]:
df['self_employed'].fillna('Not Informed', inplace=True)
df['work_interfere'].fillna('Not Informed', inplace=True)
df['state'].fillna('Not Informed', inplace=True)

In [15]:
df['self_employed'].value_counts(dropna=False)

No              1095
Yes              146
Not Informed      18
Name: self_employed, dtype: int64

In [16]:
df['work_interfere'].value_counts(dropna=False)

Sometimes       465
Not Informed    264
Never           213
Rarely          173
Often           144
Name: work_interfere, dtype: int64

In [17]:
df['state'].value_counts(dropna=False)

Not Informed    515
CA              138
WA               70
NY               57
TN               45
TX               44
OH               30
IL               29
OR               29
PA               29
IN               27
MI               22
MN               21
MA               20
FL               15
NC               14
VA               14
WI               12
GA               12
MO               12
UT               11
CO                9
MD                8
AL                8
AZ                7
OK                6
NJ                6
KY                5
SC                5
IA                4
CT                4
DC                4
NV                3
VT                3
SD                3
KS                3
NH                3
WY                2
NM                2
NE                2
WV                1
ID                1
MS                1
RI                1
LA                1
ME                1
Name: state, dtype: int64

In [18]:
df.isnull().sum()

Age                          0
Gender                       0
Country                      0
state                        0
self_employed                0
family_history               0
treatment                    0
work_interfere               0
no_employees                 0
remote_work                  0
tech_company                 0
benefits                     0
care_options                 0
wellness_program             0
seek_help                    0
anonymity                    0
leave                        0
mental_health_consequence    0
phys_health_consequence      0
coworkers                    0
supervisor                   0
mental_health_interview      0
phys_health_interview        0
mental_vs_physical           0
obs_consequence              0
dtype: int64

### Exploração Analítica (EDA)

In [19]:
categorical_features = df.columns[df.dtypes == object]
numerical_features = df.columns[df.dtypes != object]

In [20]:
categorical_features

Index(['Gender', 'Country', 'state', 'self_employed', 'family_history',
       'treatment', 'work_interfere', 'no_employees', 'remote_work',
       'tech_company', 'benefits', 'care_options', 'wellness_program',
       'seek_help', 'anonymity', 'leave', 'mental_health_consequence',
       'phys_health_consequence', 'coworkers', 'supervisor',
       'mental_health_interview', 'phys_health_interview',
       'mental_vs_physical', 'obs_consequence'],
      dtype='object')

In [21]:
numerical_features

Index(['Age'], dtype='object')

In [22]:
# Quantity of categorical features - 24
len(categorical_features)

24

In [23]:
# Unique fields in each feature

In [24]:
def unique_features(features):
    dict_features ={}
    for feature in features:
        dict_features[feature] = features[feature].value_counts()
    return dict_features

In [25]:
unique_f = unique_features(df)

In [26]:
unique_f

{'Age':  29             85
  32             82
  26             75
  27             71
  33             70
  28             68
  31             67
  34             65
  30             63
  25             61
  35             55
  23             51
  24             46
  37             43
  38             39
  36             37
  40             33
  39             33
  43             28
  22             21
  41             21
  42             20
  21             16
  45             12
  46             12
  44             11
  19              9
  18              7
  48              6
  50              6
  20              6
  51              5
  49              4
  56              4
  57              3
  54              3
  55              3
  47              2
  60              2
  99999999999     1
  5               1
 -1               1
  11              1
  8               1
  61              1
  53              1
 -29              1
 -1726            1
  65              1
  62         

### Engenharia de Features

#### 1 - Numerical Values

In [27]:
## min: -1726 and MAX: 99999999999
## get the Median and substitute
df['Age'].median()

31.0

In [28]:
# Ages > 100  => 329,99999999999
df[df['Age']>100]

Age Gender        Country         state self_employed  \
364          329   Male  United States            OH            No   
390  99999999999    All       Zimbabwe  Not Informed           Yes   

    family_history treatment work_interfere no_employees remote_work  ...  \
364             No       Yes          Often         6-25         Yes  ...   
390            Yes       Yes          Often          1-5          No  ...   

      anonymity           leave mental_health_consequence  \
364  Don't know      Don't know                     Maybe   
390          No  Very difficult                       Yes   

    phys_health_consequence     coworkers supervisor mental_health_interview  \
364                      No  Some of them         No                      No   
390                     Yes            No         No                     Yes   

    phys_health_interview mental_vs_physical obs_consequence  
364                    No                 No              No  
390                    No                 No             Yes  

[2 rows x 25 columns]

In [29]:
# Ages < 15 => -29, -1726, 5, 8, 11, -1
df[df['Age']<15]

Age              Gender         Country         state self_employed  \
143    -29                Male   United States            MN            No   
715  -1726                male  United Kingdom  Not Informed            No   
734      5                Male   United States            OH            No   
989      8  A little about you    Bahamas, The            IL           Yes   
1090    11                male   United States            OH           Yes   
1127    -1                   p   United States            AL           Yes   

     family_history treatment work_interfere    no_employees remote_work  ...  \
143              No        No   Not Informed  More than 1000         Yes  ...   
715              No       Yes      Sometimes          26-100          No  ...   
734              No        No   Not Informed         100-500          No  ...   
989             Yes       Yes          Often             1-5         Yes  ...   
1090             No        No          Never             1-5         Yes  ...   
1127            Yes       Yes          Often             1-5         Yes  ...   

       anonymity               leave mental_health_consequence  \
143   Don't know          Don't know                        No   
715   Don't know  Somewhat difficult                       Yes   
734   Don't know       Somewhat easy                        No   
989          Yes           Very easy                       Yes   
1090         Yes           Very easy                        No   
1127         Yes           Very easy                       Yes   

     phys_health_consequence     coworkers    supervisor  \
143                       No  Some of them           Yes   
715                       No            No            No   
734                       No           Yes           Yes   
989                      Yes           Yes           Yes   
1090                      No  Some of them  Some of them   
1127                     Yes           Yes           Yes   

     mental_health_interview phys_health_interview mental_vs_physical  \
143                       No                    No         Don't know   
715                       No                 Maybe         Don't know   
734                       No                    No                Yes   
989                      Yes                   Yes                Yes   
1090                      No                 Maybe                Yes   
1127                     Yes                   Yes                Yes   

     obs_consequence  
143               No  
715               No  
734               No  
989              Yes  
1090              No  
1127             Yes  

[6 rows x 25 columns]

In [30]:
# If Age > 100 or Age < 15  = Median 
df.loc[df.Age > 100, 'Age'] = df['Age'].median()
df.loc[df.Age < 15, 'Age'] = df['Age'].median()

#### 2 - Categorical Values

#### 2.1 - Gender

In [31]:
# Check quantity of categories
df['Gender'].value_counts( normalize=True ) * 100

Male                                              48.848292
male                                              16.362192
Female                                             9.610802
M                                                  9.213662
female                                             4.924543
F                                                  3.018268
m                                                  2.700556
f                                                  1.191422
Make                                               0.317712
Male                                               0.238284
Woman                                              0.238284
Cis Male                                           0.158856
Man                                                0.158856
Female (trans)                                     0.158856
Female                                             0.158856
Trans woman                                        0.079428
msle                                    

In [32]:
df['Gender'].describe()

count     1259
unique      49
top       Male
freq       615
Name: Gender, dtype: object

In [43]:
df['Gender'].unique()

array(['Female', 'M', 'Male', 'male', 'female', 'm', 'Male-ish', 'maile',
       'Trans-female', 'Cis Female', 'F', 'something kinda male?',
       'Cis Male', 'Woman', 'f', 'Mal', 'Male (CIS)', 'queer/she/they',
       'non-binary', 'Femake', 'woman', 'Make', 'Nah', 'All', 'Enby',
       'fluid', 'Genderqueer', 'Female ', 'Androgyne', 'Agender',
       'cis-female/femme', 'Guy (-ish) ^_^', 'male leaning androgynous',
       'Male ', 'Man', 'Trans woman', 'msle', 'Neuter', 'Female (trans)',
       'queer', 'Female (cis)', 'Mail', 'cis male', 'A little about you',
       'Malr', 'p', 'femail', 'Cis Man',
       'ostensibly male, unsure what that really means'], dtype=object)

In [46]:
df['Gender'] = df['Gender'].map({'Female':0, 'M':1, 'Male':1, 'male':1, 'female':0, 'm':1, 'Male-ish':2, 'maile':1,\
                                 'Trans-female':2, 'Cis Female':2, 'F':0, 'something kinda male?':2,\
                                 'Cis Male':2, 'Woman':0, 'f':0, 'Mal':1, 'Male (CIS)':2, 'queer/she/they':2,\
                                 'non-binary':2, 'Femake':0, 'woman':0, 'Make':1, 'Nah':2, 'All':2, 'Enby':2,\
                                 'fluid':2, 'Genderqueer':2, 'Female ':0, 'Androgyne':2, 'Agender':2,\
                                 'cis-female/femme':2, 'Guy (-ish) ^_^':2, 'male leaning androgynous':2,\
                                 'Male ':1, 'Man':1, 'Trans woman':2, 'msle':1, 'Neuter':2, 'Female (trans)':2,\
                                 'queer':2, 'Female (cis)':2, 'Mail':1, 'cis male':2, 'A little about you':2,\
                                 'Malr':1, 'p':2, 'femail':0, 'Cis Man':2,\
                                 'ostensibly male, unsure what that really means':2
                                })


In [47]:
df['Gender'].describe()

count    1259.000000
mean        0.830024
std         0.434616
min         0.000000
25%         1.000000
50%         1.000000
75%         1.000000
max         2.000000
Name: Gender, dtype: float64

In [35]:
df.describe(include="all")

Age Gender        Country         state self_employed  \
count   1259.000000   1259           1259          1259          1259   
unique          NaN     49             48            46             3   
top             NaN   Male  United States  Not Informed            No   
freq            NaN    615            751           515          1095   
mean      32.069897    NaN            NaN           NaN           NaN   
std        7.265565    NaN            NaN           NaN           NaN   
min       18.000000    NaN            NaN           NaN           NaN   
25%       27.000000    NaN            NaN           NaN           NaN   
50%       31.000000    NaN            NaN           NaN           NaN   
75%       36.000000    NaN            NaN           NaN           NaN   
max       72.000000    NaN            NaN           NaN           NaN   

       family_history treatment work_interfere no_employees remote_work  ...  \
count            1259      1259           1259         1259        1259  ...   
unique              2         2              5            6           2  ...   
top                No       Yes      Sometimes         6-25          No  ...   
freq              767       637            465          290         883  ...   
mean              NaN       NaN            NaN          NaN         NaN  ...   
std               NaN       NaN            NaN          NaN         NaN  ...   
min               NaN       NaN            NaN          NaN         NaN  ...   
25%               NaN       NaN            NaN          NaN         NaN  ...   
50%               NaN       NaN            NaN          NaN         NaN  ...   
75%               NaN       NaN            NaN          NaN         NaN  ...   
max               NaN       NaN            NaN          NaN         NaN  ...   

         anonymity       leave mental_health_consequence  \
count         1259        1259                      1259   
unique           3           5                         3   
top     Don't know  Don't know                        No   
freq           819         563                       490   
mean           NaN         NaN                       NaN   
std            NaN         NaN                       NaN   
min            NaN         NaN                       NaN   
25%            NaN         NaN                       NaN   
50%            NaN         NaN                       NaN   
75%            NaN         NaN                       NaN   
max            NaN         NaN                       NaN   

       phys_health_consequence     coworkers supervisor  \
count                     1259          1259       1259   
unique                       3             3          3   
top                         No  Some of them        Yes   
freq                       925           774        516   
mean                       NaN           NaN        NaN   
std                        NaN           NaN        NaN   
min                        NaN           NaN        NaN   
25%                        NaN           NaN        NaN   
50%                        NaN           NaN        NaN   
75%                        NaN           NaN        NaN   
max                        NaN           NaN        NaN   

       mental_health_interview phys_health_interview mental_vs_physical  \
count                     1259                  1259               1259   
unique                       3                     3                  3   
top                         No                 Maybe         Don't know   
freq                      1008                   557                576   
mean                       NaN                   NaN                NaN   
std                        NaN                   NaN                NaN   
min                        NaN                   NaN                NaN   
25%                        NaN                   NaN                NaN   
50%                        NaN                   NaN                NaN  

In [36]:
df[df['Gender'].isnull()]

Empty DataFrame
Columns: [Age, Gender, Country, state, self_employed, family_history, treatment, work_interfere, no_employees, remote_work, tech_company, benefits, care_options, wellness_program, seek_help, anonymity, leave, mental_health_consequence, phys_health_consequence, coworkers, supervisor, mental_health_interview, phys_health_interview, mental_vs_physical, obs_consequence]
Index: []

[0 rows x 25 columns]

#### Change Categorical values for Numeric

In [37]:
df['self_employed'] = df['self_employed'].map({'Yes':1, 'No':0})
df['family_history'] = df['family_history'].map({'Yes':1,'No':0})
df['treatment'] = df['treatment'].map({'Yes':1,'No':0})
df['remote_work'] = df['remote_work'].map({'Yes':1,'No':0})
df['tech_company'] = df['tech_company'].map({'Yes':1,'No':0})
df['obs_consequence'] = df['obs_consequence'].map({'Yes':1,'No':0})
df['work_interfere_num'] = pd.factorize(df['work_interfere'])[0]

#### More than two options in Categorical values - use pd.factorize

In [38]:
df['seek_help_num'] = pd.factorize(df['seek_help'])[0]

In [39]:
df['seek_help_num'].value_counts()

2    646
1    363
0    250
Name: seek_help_num, dtype: int64

In [40]:
#Work interfere
df[['work_interfere_num','work_interfere']].value_counts( normalize=True ) * 100

work_interfere_num  work_interfere
3                   Sometimes         36.934075
4                   Not Informed      20.969023
2                   Never             16.918189
1                   Rarely            13.741064
0                   Often             11.437649
dtype: float64

### Target: leave

In [41]:
df['treatment'].value_counts( normalize=True ) * 100

1    50.595711
0    49.404289
Name: treatment, dtype: float64

In [42]:
df['treatment'].describe()

count    1259.000000
mean        0.505957
std         0.500163
min         0.000000
25%         0.000000
50%         1.000000
75%         1.000000
max         1.000000
Name: treatment, dtype: float64